stan/math/rev/fun/fill.hpp

```diff
@@ -0,0 +1,65 @@
+#ifndef STAN_MATH_REV_FUN_FILL_HPP
+#define STAN_MATH_REV_FUN_FILL_HPP
+
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/rev/meta.hpp>
+#include <stan/math/rev/core.hpp>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+/**
+ * Fill the specified container with the specified value.
+ *
+ * The specified matrix is filled by element.
+ *
+ * @tparam VarMat a `var_value` with inner type from `EigenBase`
+ * @tparam S A var.
+ *
+ * @param x Container.
+ * @param y Value.
+ */
+template <typename VarMat, typename S, require_var_matrix_t<VarMat>* = nullptr,
+          require_var_t<S>* = nullptr>
+inline void fill(VarMat& x, const S& y) {
+  arena_t<plain_type_t<value_type_t<VarMat>>> prev_vals = x.val().eval();
+  x.vi_->val_.fill(y.val());
+  if (!is_nan(y)) {
```

Comment:
I don't like this `is_nan` behavior. I'd rather if it was a nan is just behaved like everything else and exploded. It's probably going to explode anyway.

Reply:
Yeah so I did this mostly because for `fill()` I think we use it in the stan compiler 90% of the time for filling in the matrices before we actually do stuff to them. Like the stan compiler will write stuff like

```cpp
    Eigen::Matrix<local_scalar_t__, -1, 1> out;
    out = Eigen::Matrix<local_scalar_t__, -1, 1>(size(ind_set));
    stan::math::fill(out, DUMMY_VAR__);
```
And we really don't care to do the reverse pass here since this is just initializing the values.

Maybe the right answer is to have a specific `initliaize_nan_impl()` function the compiler can use when it's just initializating the matrices?

Reply:
> Maybe the right answer is to have a specific initliaize_nan_impl() function the compiler can use when it's just initializating the matrices?

Yeah that sounds good.